### PR TITLE
Support for sharing objects across groups of clients

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

With https://github.com/senaite/senaite.core/pull/2004 becomes possible to share an object with multiple clients. However, user had to manually and individually pick the clients to share the content with.

This Pull Request:
- introduces the new content type `ClientsGroup` that makes the creation of client groups possible
- allows the user to choose client groups for objects that implement `IClientShareableBehavior` behavior. An object shared with a client group becomes accessible to contacts from all clients that are assigned to the selected group

## Current behavior before PR

Is not possible to share a given object with contacts from multiple clients in a single shot

## Desired behavior after PR is merged

Is possible to share a given object with contacts from multiple clients in a single shot

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/senaite/core/profiles/default/types/ClientsGroups.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<object name="ClientsGroups"
+        meta_type="Dexterity FTI"
+        i18n:domain="senaite.diagnosis"
```
Comment (ramonski) [inline src/senaite/core/profiles/default/types/ClientsGroups.xml:4]: typo

File: src/senaite/core/behaviors/clientshareable.py [hunk 2]
```diff
@@ -19,19 +19,129 @@
...
+from zope.annotation.interfaces import IAnnotations
+
+
+TYPES_UIDS = "senaite.core.behaviors.clientshareable.types_uids"
+
+
+class TypesAwareUIDReferenceField(UIDReferenceField):
+    """A convenient field to handle UIDs of reference objects from different
+    types (e.g. Client + ClientsGroup) properly and in a more efficient way
+    """
```
Comment (ramonski) [inline src/senaite/core/behaviors/clientshareable.py:49]: I kind of understand what you are trying to do, but my brain turned off after 10 minutes looking at the logic of this field... Too many redirections

**Review by ramonski — CHANGES_REQUESTED** [review]
I believe that using objects for the purpose of grouping is the wrong approach and it clutters the core without additional value (if no patients add-on is installed).
In my opinion I would rather take the approach with an additional `group` field (or annotation) in clients and use a keyword index.
Adding clients to groups would then happen in the listing via a modal/view action to add/remove to groups.
This would have the additional value to be easy filterable in the listing, e.g. by clicking on a group and showing only the clients that belong to the same group (similar to the labels in GitHub issues).

**Comment by xispa** [comment]
> I believe that using objects for the purpose of grouping is the wrong approach and it clutters the core without additional value (if no patients add-on is installed). In my opinion I would rather take the approach with an additional `group` field (or annotation) in clients and use a keyword index. Adding clients to groups would then happen in the listing via a modal/view action to add/remove to groups. This would have the additional value to be easy filterable in the listing, e.g. by clicking on a group and showing only the clients that belong to the same group (similar to the labels in GitHub issues).

The approach you suggest by means of a new field with a "label"-like behavior looks promising. Will give a try, thanks.

**Pull request closed without merging**